### ISSUE TITLE
[Clang] ``

### ISSUE THREAD

**Issue by philnik777** [llvm/llvm-project#100198]

**Comment by Endilll** [comment]
GCC and EDG agree that the code is valid: https://godbolt.org/z/ch5ndY4ra

**Comment by zygoloid** [comment]
Should `S` be trivially copyable? It looks to me like:
- Its copy and move assignments are deleted because of the const member
- Its copy constructor is deleted because of the user-declared move assignment
- Its move constructor is not declared because of the user-declared assignment operators
... so it looks to me like it's not copyable at all.

**Comment by AaronBallman** [comment]
References for what @zygoloid posted above:

* Copy and move assign deleted due to const member: https://eel.is/c++draft/class.copy.assign#7.1
* Deleted copy constructor: https://eel.is/c++draft/class.copy.ctor#6.sentence-2
* No move constructor declared: https://eel.is/c++draft/class.copy.ctor#8

**Comment by Endilll** [comment]
So http://eel.is/c++draft/class.copy.ctor#11 was a red herring

**Comment by AaronBallman** [comment]
Not a red herring, I think it's more that it doesn't apply because the copy constructor is deleted and there's no move constructor.

**Comment by frederick-vs-ja** [comment]
> ... so it looks to me like it's not copyable at all.

Right. Being trivially copyable doesn't imply being copyable or satisfying any other `is_trivially_*`, which is one of the most cursed fact in the C++ standard.

**Comment by frederick-vs-ja** [comment]
This seems related to [CWG1734](https://cplusplus.github.io/CWG/issues/1734.html).

**Comment by philnik777** [comment]
If that's the case, shouldn't
```
struct S {
  int offset_;
};

struct pair
{
  int first = {};
  const S second = {};

  pair() = default;
  pair(const pair&) = default;
  pair(pair&&) = default;
  pair& operator=(const pair&) = default;
  pair& operator=(pair&&) = default;
};
```
be considered trivially copyable? I can definitely copy construct it, but it's not trivially copyable according to Clang: https://godbolt.org/z/qn4K1G6sY

**Comment by AaronBallman** [comment]
We have other issues regarding calculating what is trivially copyable, e.g., https://github.com/llvm/llvm-project/issues/38398
